app/controllers/application_controller.rb

```diff
@@ -28,9 +28,16 @@ class ApplicationController < ActionController::Base
   prepend_before_action :add_new_relic_trace_attributes
   prepend_before_action :session_expires_at
   prepend_before_action :set_locale
+  prepend_before_action :set_x_request_url
   before_action :disable_caching
   before_action :cache_issuer_in_cookie
 
+  # Workaround that helps our JS fetch polyfill. See:
+  # https://www.npmjs.com/package/whatwg-fetch#user-content-obtaining-the-response-url
+  def set_x_request_url
+    response.headers['X-Request-URL'] = request.url
```

Comment:
Would this get set on non-successful responses as well? I'm having trouble testing locally, where I'm seeing some 400 responses, without the header being included, so the same original error is being logged. Not sure if the 400s are something flakey on my end, but it seems like maybe we should want to set the header regardless?

Reply:
Depending on what caused the 400, if another filter early-returned, that could cause this not to run. Do you have a repro for the 400 missing the header? If so it should be easy to test that we have the before filters in the right order

Reply:
> Depending on what caused the 400, if another filter early-returned, that could cause this not to run. Do you have a repro for the 400 missing the header? If so it should be easy to test that we have the before filters in the right order

I've tracked it down to throwing up here:

https://github.com/18F/identity-idp/blob/02d60d09ba99d9712a56c9c56063efe386c45b11/lib/utf8_sanitizer.rb#L21

Based on the network monitor, it seems to be following a redirect from the `/verify_wait` fetch response, and generating a 400 on the subsequent `GET` request. I found a similar issue at https://github.com/rack/rack/issues/903, though it seems to have been claimed to be fixed long ago.

Still not sure why it's only happening for me locally... last night I'd thought it may be because I run non-async locally normally, but it happens in both cases.

Reply:
Ok so the bad news is that's a middleware, so it's way before the controller before actions. Do you think we should move this patch to the middleware layer? Or just proceed as-is for now?

Reply:
I'm guessing it probably _shouldn't_ be hitting that `EOFError` rescue block in the first place, so my instinct is that it would be okay to not have to return the `X-Request-URL` on these responses, but rather to figure out why it's bailing there in the first place 🤔

Reply:
Tracked this down and included a fix with inline code comment in fd5309c.